pkg/machine/ocipull/ociartifact.go

```diff
@@ -0,0 +1,257 @@
+package ocipull
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/containers/image/v5/docker"
+	"github.com/containers/image/v5/docker/reference"
+	"github.com/containers/image/v5/transports/alltransports"
+	"github.com/containers/image/v5/types"
+	"github.com/containers/podman/v5/pkg/machine/compression"
+	"github.com/containers/podman/v5/pkg/machine/define"
+	"github.com/containers/podman/v5/utils"
+	"github.com/opencontainers/go-digest"
+	specV1 "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// TODO This is temporary until we decide on a proper image name
+	artifactRegistry     = "quay.io"
+	artifactRepo         = "baude"
+	artifactImageName    = "podman-machine-images-art"
+	artifactOriginalName = "org.opencontainers.image.title"
+)
+
+type OCIArtifactDisk struct {
+	cachedCompressedDiskPath *define.VMFile
+	name                     string
+	ctx                      context.Context
+	dirs                     *define.MachineDirs
+	diskArtifactOpts         *DiskArtifactOpts
+	finalPath                string
+	imageEndpoint            string
+	machineVersion           *OSVersion
+	diskArtifactFileName     string
+	pullOptions              *PullOptions
+	vmType                   define.VMType
+}
+
+type DiskArtifactOpts struct {
+	arch     string
+	diskType string
+	os       string
+}
+
+/*
+
+	This interface is for automatically pulling a disk artifact(qcow2, raw, vhdx file) from a pre-determined
+	image location.  The logic is tied to vmtypes (applehv, qemu, hyperv) and their understanding of the type of
+	disk they require.  The process can be generally described as:
+
+	* Determine the flavor of artifact we are looking for (arch, compression, type)
+	* Grab the manifest list for the target
+	* Walk the artifacts to find a match based on flavor
+	* Check the hash of the artifact against the hash of our cached image
+	* If the cached image does not exist or match, pull the latest into an OCI directory
+		* Read the OCI blob's manifest to determine which blob is the artifact disk
+		* Rename/move the blob in the OCI directory to the image cache dir and append the type and compression
+		  i.e. 91d1e51ddfac9d4afb1f96df878089cfdb9ab9be5886f8bccac0f0557ed28974.qcow2.xz
+		* Discard the OCI directory
+	* Decompress the cached image to the image dir in the form of <vmname>-<arch>.<raw|vhdx|qcow2>
+
+*/
+
+func NewOCIArtifactPull(ctx context.Context, dirs *define.MachineDirs, vmName string, vmType define.VMType, finalPath *define.VMFile) (*OCIArtifactDisk, error) {
+	var (
+		arch string
+	)
+
+	artifactVersion := getVersion()
+	switch runtime.GOARCH {
+	case "amd64":
+		arch = "x86_64"
```

Comment:
Why are you switching arch and not just using runtime.GOARCH?

Reply:
because fcos "arch" and go "arch" are different... nothing there is normalized.